community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/participant/CountsRebuilder.java

```diff
@@ -0,0 +1,170 @@
+    {
+        if ( countStoreRebuildRequired( versionToMigrateFrom ) )
+        {
+            // create counters from scratch
+            StoreFile.fileOperation( DELETE, fileSystem, migrationDir, migrationDir, COUNTS_STORE_FILES, true, null,
+                    StoreFileType.STORE );
+            File neoStore = new File( storeDir, DEFAULT_NAME );
+            long lastTxId = MetaDataStore.getRecord( pageCache, neoStore, Position.LAST_TRANSACTION_ID );
+            try
+            {
+                rebuildCountsFromScratch( migrationDir, lastTxId, progressMonitor, versionToMigrateTo, pageCache );
+            }
+            catch ( StoreFailureException e )
+            {
+                //This means that we did not perform a full migration, as the formats had the same capabilities. Thus
+                // we should use the store directory for information when rebuilding the count store.
+                rebuildCountsFromScratch( storeDir, lastTxId, progressMonitor, versionToMigrateFrom, pageCache );
+            }
+        }
+        progressMonitor.completed();
+    }
+
+    @Override
+    public void moveMigratedFiles( File migrationDir, File storeDir, String versionToUpgradeFrom,
+            String versionToUpgradeTo ) throws IOException
+    {
+        // Move the migrated ones into the store directory
+        StoreFile.fileOperation( MOVE, fileSystem, migrationDir, storeDir, COUNTS_STORE_FILES, true,
+                // allow to skip non existent source files
+                ExistingTargetStrategy.OVERWRITE, // allow to overwrite target files
+                StoreFileType.values() );
+        // We do not need to move files with the page cache, as the count files always reside on the normal file system.
+    }
+
+    @Override
+    public void cleanup( File migrationDir ) throws IOException
+    {
+        fileSystem.deleteRecursively( migrationDir );
+    }
+
+    @Override
+    public String toString()
+    {
+        return "Kernel Node Count Rebuilder";
+    }
+
+    boolean countStoreRebuildRequired( String versionToMigrateFrom )
+    {
+        return StandardV2_3.STORE_VERSION.equals( versionToMigrateFrom ) ||
+               StandardV3_0.STORE_VERSION.equals( versionToMigrateFrom ) ||
+               StoreVersion.HIGH_LIMIT_V3_0_0.versionString().equals( versionToMigrateFrom ) ||
+               StoreVersion.HIGH_LIMIT_V3_0_6.versionString().equals( versionToMigrateFrom ) ||
+               StoreVersion.HIGH_LIMIT_V3_1_0.versionString().equals( versionToMigrateFrom );
+    }
+
+    private void rebuildCountsFromScratch( File migrationDir, long lastTxId,
+            MigrationProgressMonitor.Section progressMonitor, String expectedStoreVersion, PageCache pageCache )
+    {
+        final File storeFileBase = new File( migrationDir, MetaDataStore.DEFAULT_NAME + StoreFactory.COUNTS_STORE );
+
+        RecordFormats recordFormats = selectForVersion( expectedStoreVersion );
```

Comment:
as the result of running this migration user will see following:
```
2017-06-20 12:03:29.516+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] Selected RecordFormat:StandardV3_0[v0.A.7] record format from store /data/databases/graph.db
2017-06-20 12:03:29.516+0000 INFO [o.n.k.i.s.f.RecordFormatSelector] Selected format 'RecordFormat:StandardV3_2[v0.A.8]' for existing store with format 'RecordFormat:StandardV3_0[v0.A.7]'
2017-06-20 12:03:29.534+0000 INFO [o.n.k.i.s.p.StoreMigrator] Starting upgrade of database
2017-06-20 12:03:29.564+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Indexes (1/4):
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-06-20 12:03:29.567+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-06-20 12:03:29.568+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-06-20 12:03:29.568+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-06-20 12:03:29.568+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-06-20 12:03:29.568+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-06-20 12:03:29.568+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Legacy indexes (2/4):
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-06-20 12:03:29.569+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-06-20 12:03:29.570+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-06-20 12:03:29.570+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-06-20 12:03:29.570+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Store files (3/4):
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-06-20 12:03:29.583+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-06-20 12:03:29.584+0000 INFO [o.n.k.i.s.p.StoreMigrator] Migrating Count rebuilding (4/4):
2017-06-20 12:03:29.647+0000 ERROR [o.n.k.i.s.c.CountsTracker] Failed to open counts store file: /graph.db/neostore.counts.db.a Format header/trailer has changed. Expected format:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x02, 'V']`, actual:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x01, 'V']`.
java.lang.IllegalStateException: Format header/trailer has changed. Expected format:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x02, 'V']`, actual:`['N', 'e', 'o', 'C', 'o', 'u', 'n', 't', 'S', 't', 'o', 'r', 'e', 0x00, 0x01, 'V']`.
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State.readFormatSpecifier(MetadataCollector.java:242)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State.access$100(MetadataCollector.java:114)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector$State$1.visit(MetadataCollector.java:121)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector.visit(MetadataCollector.java:69)
	at org.neo4j.kernel.impl.store.kvstore.MetadataCollector.visit(MetadataCollector.java:29)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFile.visitKeyValuePairs(KeyValueStoreFile.java:192)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFile.scanAll(KeyValueStoreFile.java:183)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFileFormat.open(KeyValueStoreFileFormat.java:287)
	at org.neo4j.kernel.impl.store.kvstore.KeyValueStoreFileFormat.openStore(KeyValueStoreFileFormat.java:78)
	at org.neo4j.kernel.impl.store.kvstore.RotationStrategy.open(RotationStrategy.java:65)
	at org.neo4j.kernel.impl.store.kvstore.DeadState$Stopped.initialize(DeadState.java:125)
	at org.neo4j.kernel.impl.store.kvstore.AbstractKeyValueStore.init(AbstractKeyValueStore.java:172)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:406)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:435)
	at org.neo4j.kernel.lifecycle.LifeSupport.bringToState(LifeSupport.java:315)
	at org.neo4j.kernel.lifecycle.LifeSupport.add(LifeSupport.java:248)
	at org.neo4j.kernel.lifecycle.Lifespan.add(Lifespan.java:41)
	at org.neo4j.kernel.impl.storemigration.participant.CountsRebuilder.rebuildCountsFromScratch(CountsRebuilder.java:165)
	at org.neo4j.kernel.impl.storemigration.participant.CountsRebuilder.migrate(CountsRebuilder.java:106)
	at org.neo4j.kernel.impl.storemigration.StoreUpgrader.migrateToIsolatedDirectory(StoreUpgrader.java:210)
	at org.neo4j.kernel.impl.storemigration.StoreUpgrader.migrateIfNeeded(StoreUpgrader.java:134)
	at org.neo4j.kernel.impl.storemigration.DatabaseMigrator.migrate(DatabaseMigrator.java:100)
	at org.neo4j.kernel.NeoStoreDataSource.upgradeStore(NeoStoreDataSource.java:568)
	at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:430)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:100)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:205)
	at org.neo4j.graphdb.EnterpriseGraphDatabase.<init>(EnterpriseGraphDatabase.java:44)
	at org.neo4j.server.enterprise.EnterpriseNeoServer.lambda$static$1(EnterpriseNeoServer.java:74)
	at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:89)
	at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:442)
	at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
	at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:211)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:107)
	at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:80)
	at org.neo4j.server.enterprise.EnterpriseEntryPoint.main(EnterpriseEntryPoint.java:36)
2017-06-20 12:03:29.704+0000 INFO [o.n.k.i.s.p.StoreMigrator]   10% completed
2017-06-20 12:03:29.704+0000 INFO [o.n.k.i.s.p.StoreMigrator]   20% completed
2017-06-20 12:03:29.704+0000 INFO [o.n.k.i.s.p.StoreMigrator]   30% completed
2017-06-20 12:03:29.704+0000 INFO [o.n.k.i.s.p.StoreMigrator]   40% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   50% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   60% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   70% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   80% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   90% completed
2017-06-20 12:03:29.737+0000 INFO [o.n.k.i.s.p.StoreMigrator]   100% completed
2017-06-20 12:03:29.798+0000 INFO [o.n.k.i.s.p.StoreMigrator] Successfully finished upgrade of database
```
i do not think that user should see an exception during performed migration